filter.go

```diff
@@ -49,25 +53,69 @@ func parseUsernameAndPassword(auth string) (username, password string, ok bool)
 	return username, password, true
 }
 
-func dial(config *config) (*ldap.Conn, error) {
+func Connect(conf *config) (*ldap.Conn, error) {
+	var certPool *x509.CertPool
+
+	if conf.certificateAuthority != "" {
+		certPool = x509.NewCertPool()
+		certPool.AppendCertsFromPEM([]byte(conf.certificateAuthority))
+	}
+
+	var conn *ldap.Conn = nil
+	var err error = nil
+	if conf.tls {
+		tlsCfg := &tls.Config{
+			InsecureSkipVerify: conf.insecureSkipVerify,
+			ServerName:         conf.host,
+			RootCAs:            certPool,
+		}
+		if conf.startTLS {
+			conn, err = dial(conf)
+			if err == nil {
+				err = conn.StartTLS(tlsCfg)
+			}
+		} else {
+			conn, err = dialTLS(conf, tlsCfg)
+		}
+	} else {
+		conn, err = dial(conf)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
+
+func dialTLS(conf *config, tlsCfg *tls.Config) (*ldap.Conn, error) {
+	return ldap.DialURL(
+		fmt.Sprintf("ldaps://%s:%d", conf.host, conf.port),
+		ldap.DialWithTLSDialer(
+			tlsCfg,
+			&net.Dialer{
+				Timeout: time.Duration(conf.timeout) * time.Second,
+			}),
+	)
+}
+
+func dial(conf *config) (*ldap.Conn, error) {
 	return ldap.DialURL(
-		// TODO: support TLS
-		fmt.Sprintf("ldap://%s:%d", config.host, config.port),
+		fmt.Sprintf("ldap://%s:%d", conf.host, conf.port),
```

Comment:
use the log api?

Reply:
I did not comprehend how this should be handled. Which logging API can be used to achieve string formatting output similar to `fmt.Sprintf`?

Reply:
oh, sorry, my mistake.